File: automator/src/main/java/com/stardust/automator/filter/BooleanFilter.java
package com.stardust.automator.filter;

import android.os.Build;
import android.support.annotation.RequiresApi;

import com.stardust.automator.UiObject;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Stardust on 2017/3/9.
 */

public class BooleanFilter extends DfsFilter {

    private static Map<BooleanSupplier, BooleanFilter[]> cache = new HashMap<>();


    public static BooleanFilter get(BooleanSupplier supplier, boolean b) {
        BooleanFilter[] booleanFilters = cache.get(supplier);
        if (booleanFilters == null) {
            booleanFilters = new BooleanFilter[2];
            cache.put(supplier, booleanFilters);
        }
        int i = b ? 1 : 0;
        if (booleanFilters[i] == null) {
            booleanFilters[i] = new BooleanFilter(supplier, b);
        }
        return booleanFilters[i];
    }

    public interface BooleanSupplier {

        boolean get(UiObject node);

    }

    public static final BooleanSupplier CHECKABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isCheckable();
        }

        @Override
        public String toString() {
            return "checkable";
        }
    };

    public static final BooleanSupplier CHECKED = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isChecked();
        }

        @Override
        public String toString() {
            return "checked";
        }
    };

    public static final BooleanSupplier FOCUSABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isFocusable();
        }

        @Override
        public String toString() {
            return "focusable";
        }
    };

    public static final BooleanSupplier FOCUSED = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isFocused();
        }

        @Override
        public String toString() {
            return "focused";
        }
    };

    public static final BooleanSupplier VISIBLE_TO_USER = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isVisibleToUser();
        }

        @Override
        public String toString() {
            return "visibleToUser";
        }
    };

    public static final BooleanSupplier ACCESSIBILITY_FOCUSED = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isAccessibilityFocused();
        }

        @Override
        public String toString() {
            return "accessibilityFocused";
        }
    };

    public static final BooleanSupplier SELECTED = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isSelected();
        }

        @Override
        public String toString() {
            return "selected";
        }
    };

    public static final BooleanSupplier CLICKABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isClickable();
        }

        @Override
        public String toString() {
            return "clickable";
        }
    };

    public static final BooleanSupplier LONG_CLICKABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isLongClickable();
        }

        @Override
        public String toString() {
            return "longClickable";
        }
    };

    public static final BooleanSupplier ENABLED = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isEnabled();
        }

        @Override
        public String toString() {
            return "enabled";
        }
    };


    public static final BooleanSupplier PASSWORD = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isPassword();
        }

        @Override
        public String toString() {
            return "password";
        }
    };

    public static final BooleanSupplier SCROLLABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isScrollable();
        }

        @Override
        public String toString() {
            return "scrollable";
        }
    };

    public static final BooleanSupplier EDITABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isEditable();
        }

        @Override
        public String toString() {
            return "editable";
        }
    };
    public static final BooleanSupplier CONTENT_INVALID = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isContentInvalid();
        }

        @Override
        public String toString() {
            return "contentInvalid";
        }
    };

    public static final BooleanSupplier CONTEXT_CLICKABLE = new BooleanSupplier() {

        @RequiresApi(api = Build.VERSION_CODES.M)
        @Override
        public boolean get(UiObject node) {
            return node.isContextClickable();
        }

        @Override
        public String toString() {
            return "checkable";
        }
    };

    public static final BooleanSupplier MULTI_LINE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isMultiLine();
        }

        @Override
        public String toString() {
            return "multiLine";
        }
    };

    public static final BooleanSupplier DISMISSABLE = new BooleanSupplier() {

        @Override
        public boolean get(UiObject node) {
            return node.isDismissable();
        }

        @Override
        public String toString() {
            return "dismissable";
        }
    };

    private BooleanSupplier mBooleanSupplier;
    private boolean mExceptedValue;

    public BooleanFilter(BooleanSupplier booleanSupplier, boolean exceptedValue) {
        mBooleanSupplier = booleanSupplier;
        mExceptedValue = exceptedValue;
    }

    @Override
    protected boolean isIncluded(UiObject nodeInfo) {
        return nodeInfo != null && mBooleanSupplier.get(nodeInfo) == mExceptedValue;
    }

    @Override
    public String toString() {
        return mBooleanSupplier.toString() + "(" + mExceptedValue + ")";
    }
}

File: automator/src/main/java/com/stardust/automator/filter/BooleanFilter.kt
package com.stardust.automator.filter

import android.os.Build
import androidx.annotation.RequiresApi

import com.stardust.automator.UiObject

import java.util.HashMap

/**
 * Created by Stardust on 2017/3/9.
 */

class BooleanFilter(private val mBooleanSupplier: BooleanSupplier, private val mExceptedValue: Boolean) : Filter {

    interface BooleanSupplier {

        operator fun get(node: UiObject): Boolean

    }

    override fun filter(node: UiObject): Boolean {
        return mBooleanSupplier[node] == mExceptedValue
    }

    override fun toString(): String {
        return mBooleanSupplier.toString() + "(" + mExceptedValue + ")"
    }

    companion object {

        private val cache = HashMap<BooleanSupplier, Array<BooleanFilter>>()

        operator fun get(supplier: BooleanSupplier, b: Boolean): BooleanFilter {
            var booleanFilters: Array<BooleanFilter>? = cache[supplier]
            if (booleanFilters == null) {
                booleanFilters = Array(2) {
                    BooleanFilter(supplier, it != 0)
                }
                cache[supplier] = booleanFilters
            }
            val i = if (b) 1 else 0
            return booleanFilters[i]
        }

        val CHECKABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isCheckable
            }

            override fun toString(): String {
                return "checkable"
            }
        }

        val CHECKED: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isChecked
            }

            override fun toString(): String {
                return "checked"
            }
        }

        val FOCUSABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isFocusable
            }

            override fun toString(): String {
                return "focusable"
            }
        }

        val FOCUSED: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isFocused
            }

            override fun toString(): String {
                return "focused"
            }
        }

        val VISIBLE_TO_USER: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isVisibleToUser
            }

            override fun toString(): String {
                return "visibleToUser"
            }
        }

        val ACCESSIBILITY_FOCUSED: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isAccessibilityFocused
            }

            override fun toString(): String {
                return "accessibilityFocused"
            }
        }

        val SELECTED: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isSelected
            }

            override fun toString(): String {
                return "selected"
            }
        }

        val CLICKABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isClickable
            }

            override fun toString(): String {
                return "clickable"
            }
        }

        val LONG_CLICKABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isLongClickable
            }

            override fun toString(): String {
                return "longClickable"
            }
        }

        val ENABLED: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isEnabled
            }

            override fun toString(): String {
                return "enabled"
            }
        }


        val PASSWORD: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isPassword
            }

            override fun toString(): String {
                return "password"
            }
        }

        val SCROLLABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isScrollable
            }

            override fun toString(): String {
                return "scrollable"
            }
        }

        val EDITABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isEditable
            }

            override fun toString(): String {
                return "editable"
            }
        }
        val CONTENT_INVALID: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isContentInvalid
            }

            override fun toString(): String {
                return "contentInvalid"
            }
        }

        val CONTEXT_CLICKABLE: BooleanSupplier = object : BooleanSupplier {

            @RequiresApi(api = Build.VERSION_CODES.M)
            override fun get(node: UiObject): Boolean {
                return node.isContextClickable
            }

            override fun toString(): String {
                return "checkable"
            }
        }

        val MULTI_LINE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isMultiLine
            }

            override fun toString(): String {
                return "multiLine"
            }
        }

        val DISMISSABLE: BooleanSupplier = object : BooleanSupplier {

            override fun get(node: UiObject): Boolean {
                return node.isDismissable
            }

            override fun toString(): String {
                return "dismissable"
            }
        }
    }
}
